### DISCUSSION TITLE
Issues with multiple providers

### DISCUSSION THREAD

**Discussion by gabsdotco** [chakra-ui/chakra-ui#5180] [Q&A]
Today I work in a project that provide two ChakraProvider components, one for the base theme of the application and other for custom themes created by the user, like the example bellow:

```tsx
import baseTheme from "theme";

export const ExampleComponent = ({ userTheme }) => {
  return (
    <ChakraProvider theme={baseTheme}>
      <Flex background="primary.500">Styled by the base theme</Flex>
      <ChakraProvider theme={userTheme}>
        <Flex background="primary.500">Styled by the user theme</Flex>
      </ChakraProvider>
    </ChakraProvider>
  )
}
```

The main goal with that structure is to isolate user styles from the base theme styles. The problem is, with that structure the user styles are overwriting the base theme styles for some reason. I don't know if the provider with the user theme is overwriting the base theme provider state or something. Any ideas?

**Comment by gabsdotco** [comment]
Update:

During a pair programming we found that the problem is related to the chakra css injection on the HTML root, so, regardless of provider level, the css injection is made on top of the application instead of the provider scope. The only workaround we found is to don't create theme tokens with same name. If anyone have a different solution, we appreciate it.

**Reply by primos63** [reply]
I created a (ChakraNestedProvider)[https://codesandbox.io/s/chakra-ui-theme-extension-w5u2n?file=/src/components/ChakraNestedProvider/index.js] that I use in my CodeSandbox so that I can have specific themes for my examples.
I use it for each `Section` like this
```jsx
export const Section = ({ children, heading, description, theme, ...rest }) => {
  const defaultTheme = useTheme()
  const sectionTheme = theme ? theme : defaultTheme

  const id = 'section-id-' + Date.now()

  return (
    <Box
      id={id}
      borderColor={useColorModeValue('blue.700', 'blue.200')}
      borderWidth={1}
      mt={5}
      borderRadius={10}
      {...rest}
      as='section'
      w='100%'
    >
      <ChakraNestedProvider theme={sectionTheme} cssVarsRoot={'#' + id}>
        <Flex
          direction='column'
          alignItems='center'
          justifyContent='center'
          mt={3}
        >
          <Heading as='h3' fontSize={18}>
            {heading}
          </Heading>
          <Text>{description}</Text>
        </Flex>
        {children}
      </ChakraNestedProvider>
    </Box>
  )
}
```

**Reply by phil-hodgson-1st** [reply]
Thank you @primos63 your answer really helped with a similar multi-theme aspect we're using in our menu system. We found that just using the `cssVarsRoot` prop seemed to override the nested components correctly with the `ChakraProvider` out of the box.

```
import { ChakraProvider } from '@chakra-ui/react';
import { customTheme2 } from '@ui';

...
<Box
  id={subMenuId}
>
  <ChakraProvider theme={customTheme2} cssVarsRoot={'#' + subMenuId}>
    <SubMenu />
  </ChakraProvider>
</Box>

**Reply by primos63** [reply]
Yes, but as I recall changing color modes will be an issue. If you don't change modes then you probably won't run into issues.

**Reply by phil-hodgson-1st** [reply]
Ah great shout! Yes we don't (currently) worry about the color modes. Thanks for the feedback 👍

**Comment by monkfromearth** [comment]
Thanks @primos63 for `ChakraNestedProvider`.

When it did not seem to work for generic variables such as `colors` in the theme.
We ended up using `body` as the `cssVarRoot` in the Global `ChakraProvider`, which seemed to make it all work.

A proof of concept for the same here: https://codesandbox.io/s/chakra-ui-theme-extension-forked-shjspw?file=/src/index.js

**Comment by trentonsnyder** [comment]
The only thing I haven't found a workaround for is `--chakra-fonts-body` while trying to use `extendTheme`. Chakra sets `--chakra-fonts-body` as a global on `body` regardless of `cssVarsRoot`. Then components are all set by default to inherit their font from it, so overriding it would require a change to all components to get their `font-family` from the var explicitly and not just `body`. So any change to this var in a nested provider theme will actually propagate up to the outer provider.

Otherwise, the nested provider suggestion here works great. Thanks everyone here for your suggestions 👍

**Reply by tae100k** [reply]
This is so weird.
@trentonsnyder is right. 
Everything works fine but font color

**Comment by salemhilal** [comment]
It seems like the variable scope is now `cssVarsRoot`, _plus_ `, [data-theme]`. This means that when nested providers are used, the parent _and_ the child's CSS variables apply to `$root, [data-theme]`. In other words, anything with the `data-theme` attribute will get whatever variables were loaded most recently, rather than whatever variables are within its scope. I think that was introduced here:
https://github.com/chakra-ui/chakra-ui/pull/5532

**Reply by stephenasuncionDEV** [reply]
@segunadebayo Please take a look at this.
